### PR TITLE
feat(spanner): add admin instance wrapper

### PR DESCRIPTION
**Pull request by bajajneha27**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:

- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/google-cloud-ruby/issues) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea.
- [x] Follow the instructions in [CONTRIBUTING](CONTRIBUTING.md). Most importantly, ensure the tests and linter pass by running `bundle exec rake ci` in the gem subdirectory.
- [x] Update code documentation if necessary.

### CODE REVIEW

**Comment by hengfengli** [comment]
@bajajneha27 Can you please fix the tests before we can review this PR? Thanks.

**Comment by bajajneha27** [comment]
> @bajajneha27 Can you please fix the tests before we can review this PR? Thanks.

Yeah, I'm working on it. I'll update once done. Thanks.

**Comment by bajajneha27** [comment]
Hi @hengfengli , does emulator support update instance request? I'm getting the following error [here](https://github.com/googleapis/google-cloud-ruby/runs/4286550390?check_suite_focus=true): 
```
Google::Cloud::UnimplementedError: 12:Cloud Spanner Emulator does not support updating instances
```

**Comment by hengfengli** [comment]
> Hi @hengfengli , does emulator support update instance request? I'm getting the following error [here](https://github.com/googleapis/google-cloud-ruby/runs/4286550390?check_suite_focus=true):
> 
> ```
> Google::Cloud::UnimplementedError: 12:Cloud Spanner Emulator does not support updating instances
> ```

No I don't think it supports that. You have to skip the test just for the emulator.

File: google-cloud-spanner/test/google/cloud/spanner/admin/instance/client_test.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+class Google::Cloud::Spanner::AdminInstanceClientConstructionMinitest < Minitest::Test
+  def test_instance_admin
+    emulator_host = "localhost:4567"
+    emulator_check = ->(name) { (name == "SPANNER_EMULATOR_HOST") ? emulator_host : nil }
+    # Clear all environment variables, except SPANNER_EMULATOR_HOST
+    ENV.stub :[], emulator_check do
+      Gapic::ServiceStub.stub :new, :stub do
+          client = Google::Cloud::Spanner::Admin::Instance.instance_admin project_id: "1234"
+          assert_kind_of Google::Cloud::Spanner::Admin::Instance::V1::InstanceAdmin::Client, client
+          assert_equal emulator_host, client.configure.endpoint
```
Comment (skuruppu) [inline google-cloud-spanner/test/google/cloud/spanner/admin/instance/client_test.rb:30]: In https://github.com/googleapis/google-cloud-ruby/pull/13456, I was also trying to assert if the credentials are `:this_channel_is_insecure`. But I couldn't get it working. Would you be able to figure out how to get it working?

**Review by skuruppu — APPROVED** [review]

File: google-cloud-spanner/lib/google/cloud/spanner/admin/instance.rb [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+            config_attr :lib_version,   nil, ::String,  nil
+            config_attr :interceptors,  nil, ::Array,   nil
+            config_attr :timeout,       nil, ::Numeric, nil
+            config_attr :quota_project, nil, ::String,  nil
+            config_attr :emulator_host, nil, ::String,  nil
+            config_attr :query_options, nil, ::Hash,    nil
+            config_attr :metadata,      nil, ::Hash,    nil
+            config_attr :retry_policy,  nil, ::Hash,    nil
+
+            # @private
```
Comment (hengfengli) [inline google-cloud-spanner/lib/google/cloud/spanner/admin/instance.rb:310]: remove?

**Review by hengfengli — APPROVED** [review]
LGTM.

File: google-cloud-spanner/test/google/cloud/spanner/admin/instance/client_test.rb (hunk 1 above)
Comment (bajajneha27) [reply google-cloud-spanner/test/google/cloud/spanner/admin/instance/client_test.rb:30]: Thanks Shanika for pointing this out. The `credentials` is actually set to `:this_channel_is_insecure` at [L105](https://github.com/googleapis/google-cloud-ruby/pull/13456/files#diff-5a08f7db57758b2198fe4b1ebaa36dc9a71ba4fd082dfffa81486c2f8d889b72R105) but then it gets changed over [here](https://github.com/googleapis/google-cloud-ruby/pull/13456/files#diff-5a08f7db57758b2198fe4b1ebaa36dc9a71ba4fd082dfffa81486c2f8d889b72R121)
I'm not sure about the right behaviour here. Can you please help me out?

Comment (quartzmo) [reply google-cloud-spanner/test/google/cloud/spanner/admin/instance/client_test.rb:30]: @bajajneha27 @skuruppu Unfortunately the actual [`GRPC::Core::Channel`]() object does not expose the provided `credentials`, afaik. However, you can use `stub` to intercept and test the arguments provided to `GRPC::Core::Channel#new`. I will provide a GH suggestion.

File: google-cloud-spanner/test/google/cloud/spanner/admin/instance/client_test.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+# limitations under the License.
+
+require "helper"
+require "google/cloud/spanner/admin/instance"
+require "google/cloud/spanner/admin/instance/v1/instance_admin"
+require "gapic/common"
+require "gapic/grpc"
+
+class Google::Cloud::Spanner::AdminInstanceClientConstructionMinitest < Minitest::Test
+  def test_instance_admin
```
Comment (quartzmo) [inline google-cloud-spanner/test/google/cloud/spanner/admin/instance/client_test.rb:22]:
```suggestion
  def test_instance_admin
    emulator_host = "localhost:4567"
    channel_args = { "grpc.service_config_disable_resolution" => 1 }
    channel_creds = :this_channel_is_insecure
    channel = GRPC::Core::Channel.new emulator_host, channel_args, channel_creds
    emulator_check = ->(name) { (name == "SPANNER_EMULATOR_HOST") ? emulator_host : nil }
    channel_check = ->(chan_host, chan_args, chan_creds) do
      assert_equal emulator_host, chan_host
      assert_equal channel_args, chan_args
      assert_equal channel_creds, chan_creds
      channel
    end
    # Clear all environment variables, except SPANNER_EMULATOR_HOST
    ENV.stub :[], emulator_check do
      Gapic::ServiceStub.stub :new, :stub do
        GRPC::Core::Channel.stub :new, channel_check do
          client = Google::Cloud::Spanner::Admin::Instance.instance_admin project_id: "1234"
          assert_kind_of Google::Cloud::Spanner::Admin::Instance::V1::InstanceAdmin::Client, client
          assert_equal emulator_host, client.configure.endpoint
        end
      end
    end
  end
```

File: google-cloud-spanner/test/google/cloud/spanner/admin/instance/client_test.rb (hunk 1 above)
Comment (quartzmo) [reply google-cloud-spanner/test/google/cloud/spanner/admin/instance/client_test.rb:30]: See suggested change, above.

**Review by quartzmo — APPROVED** [review]
LGTM. I added a suggested change in response to @skuruppu's comment on `client_test.rb:30`.

File: google-cloud-spanner/test/google/cloud/spanner/admin/instance/client_test.rb (hunk 1 above)
Comment (bajajneha27) [reply google-cloud-spanner/test/google/cloud/spanner/admin/instance/client_test.rb:30]: Thank you @quartzmo

**Pull request merged**